cli/src/main/java/org/aya/cli/literate/Highlighter.java

```diff
@@ -0,0 +1,242 @@
+
+  /// region Var
+
+  private void highlightVarDef(@NotNull AnyVar var, @Nullable HighlightInfoType.DefKind kind) {
+    var sourcePos = switch (var) {
+      case GeneralizedVar genVar -> genVar.sourcePos;
+      case DefVar<?, ?> defVar -> {
+        var concrete = defVar.concrete;
+        assert concrete != null;
+
+        yield concrete.sourcePos();
+      }
+      case LocalVar localVar -> localVar.definition();
+      default -> throw new UnsupportedOperationException("Unexpected var: " + var.getClass());
+    };
+
+    linkDef(sourcePos, var, kind);
+  }
+
+  private void highlightVarRef(@NotNull SourcePos sourcePos, @NotNull AnyVar var) {
+    highlightVarRef(sourcePos, var, kindOf(var));
+  }
+
+  private void highlightVarRef(@NotNull SourcePos sourcePos, @NotNull AnyVar var, @Nullable HighlightInfoType.DefKind kind) {
+    linkRef(sourcePos, var, kind);
+  }
+
+  /**
+   * @see org.aya.pretty.doc.Doc#linkDef(Doc, int)
+   */
+  private void linkDef(@NotNull SourcePos sourcePos, @NotNull AnyVar var, @Nullable HighlightInfoType.DefKind style) {
+    addInfo(sourcePos, new HighlightInfoType.Def(
+      String.valueOf(var.hashCode()),
+      style
+    ));
+  }
+
+  private void linkRef(@NotNull SourcePos sourcePos, @NotNull AnyVar var, @Nullable HighlightInfoType.DefKind style) {
+    addInfo(sourcePos, new HighlightInfoType.Ref(
+      String.valueOf(var.hashCode()),
+      style
+    ));
+  }
+
+  /// endregion
+
+  /// region Expr
+
+  @Override
+  @SuppressWarnings("unused")
+  public @NotNull Expr pre(@NotNull Expr expr) {
+    switch (expr) {
+      // leaves
+      case Expr.Ref ref -> highlightVarRef(ref.sourcePos(), ref.resolvedVar());
+      case Expr.RawSort rawSort -> highlightSort(rawSort.sourcePos());
+      case Expr.Sort sort -> {
+        var sourcePos = switch (sort) {
+          case Expr.ISet iSet -> iSet.sourcePos();
+          case Expr.Prop prop -> prop.sourcePos();
+          // Also highlight the number after `Type/Set`
+          case Expr.Set set -> set.sourcePos();
+          case Expr.Type type -> type.sourcePos();
+        };
+
+        highlightSort(sourcePos);
+      }
+      case Expr.LitInt litInt -> addInfo(litInt.sourcePos(), HighlightInfoType.LitInt.INSTANCE);
+      case Expr.LitString litString -> addInfo(litString.sourcePos(), HighlightInfoType.LitString.INSTANCE);
+      case Expr.Error error -> highlightError(error.sourcePos(), error.description().toDoc(options));
+      case Expr.Unresolved unresolved -> highlightError(unresolved.sourcePos(), null);
+      // TODO: how to impl?
+      case Expr.Lift lift -> {}
+      case Expr.Proj proj -> {}
+      case Expr.RawProj rawProj -> {}
+      // only highlight keyword (but this will done in another place)
+      case Expr.Coe coe -> {}
+      case Expr.Do aDo -> {}
+      case Expr.Pi pi -> {}
+      case Expr.New aNew -> {}
+      case Expr.Match match -> {}
+      // no-op
+      case Expr.Array array -> {}
+      case Expr.Idiom idiom -> {}
```

Comment:
Maybe a `case Expr misc -> {}` ?

Reply:
I hope that the compiler will remind you to impl the new case if you create a new subclass of Expr.